common/nymnoise/src/stream.rs

```diff
@@ -0,0 +1,224 @@
+// Copyright 2025 - Nym Technologies SA <contact@nymtech.net>
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{config::NoiseConfig, error::NoiseError};
+use bytes::BytesMut;
+use futures::{Sink, SinkExt, Stream, StreamExt};
+use nym_crypto::asymmetric::x25519;
+use pin_project::pin_project;
+use snow::{Builder, HandshakeState, TransportState};
+use std::io;
+use std::pin::Pin;
+use std::task::Poll;
+use std::{cmp::min, task::ready};
+use tokio::{
+    io::{AsyncRead, AsyncWrite, ReadBuf},
+    net::TcpStream,
+};
+use tokio_util::codec::{Framed, LengthDelimitedCodec};
+
+const TAGLEN: usize = 16;
+const HANDSHAKE_MAX_LEN: usize = 1024; // using this constant to limit the handshake's buffer size
+
+pub(crate) type Psk = [u8; 32];
+
+/// Wrapper around a TcpStream
+#[pin_project]
+pub struct NoiseStream {
+    #[pin]
+    inner_stream: Framed<TcpStream, LengthDelimitedCodec>,
+    handshake: Option<HandshakeState>,
+    noise: Option<TransportState>,
+    dec_buffer: BytesMut,
+}
+
+impl NoiseStream {
+    pub(crate) fn new_initiator(
+        inner_stream: TcpStream,
+        config: &NoiseConfig,
+        remote_pub_key: &x25519::PublicKey,
+        psk: &Psk,
+    ) -> Result<NoiseStream, NoiseError> {
+        let handshake = Builder::new(config.pattern.as_noise_params())
+            .local_private_key(config.local_key.private_key().as_bytes())
+            .remote_public_key(&remote_pub_key.to_bytes())
+            .psk(config.pattern.psk_position(), psk)
+            .build_initiator()?;
+        Ok(NoiseStream::new_inner(inner_stream, handshake))
+    }
+
+    pub(crate) fn new_responder(
+        inner_stream: TcpStream,
+        config: &NoiseConfig,
+        psk: &Psk,
+    ) -> Result<NoiseStream, NoiseError> {
+        let handshake = Builder::new(config.pattern.as_noise_params())
+            .local_private_key(config.local_key.private_key().as_bytes())
+            .psk(config.pattern.psk_position(), psk)
+            .build_responder()?;
+        Ok(NoiseStream::new_inner(inner_stream, handshake))
+    }
+
+    fn new_inner(inner_stream: TcpStream, handshake: HandshakeState) -> NoiseStream {
+        NoiseStream {
+            inner_stream: LengthDelimitedCodec::builder()
+                .length_field_type::<u16>()
+                .new_framed(inner_stream),
+            handshake: Some(handshake),
+            noise: None,
+            dec_buffer: BytesMut::new(),
+        }
+    }
+
+    pub(crate) async fn perform_handshake(mut self) -> Result<Self, NoiseError> {
+        //Check if we are in the correct state
+        let Some(mut handshake) = self.handshake.take() else {
+            return Err(NoiseError::IncorrectStateError);
+        };
+
+        while !handshake.is_handshake_finished() {
+            if handshake.is_my_turn() {
+                self.send_handshake_msg(&mut handshake).await?;
```

Comment:
looking slightly deeper into it. i'm a bit confused: why are you having a method on `self` that takes `handshake` argument that originally was part of self but you temporarily removed it?

Reply:
We still need the `inner_stream` from `self` to send the encrypted handshake message.
I'm taking the `handshake` out from `self` to prevent duplicate calls to this `fn` from messing with the first one.